EditingStyle: Avoid unchecked local variables

vitorroriz opened the pull request:
#### 2d4b420bfd1804cdd34c10a965f0ec01c9b80e50
<pre>
EditingStyle: Avoid unchecked local variables
<a href="https://rdar.apple.com/157719495">rdar://157719495</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297036">https://bugs.webkit.org/show_bug.cgi?id=297036</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::textDirectionForSelection):

Canonical link: <a href="https://commits.webkit.org/298372@main">https://commits.webkit.org/298372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/838f36b5151a5a785e49f50068382a52a2b3c5f6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121416 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/65912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6c06a102-c44a-4273-8fce-952881bfa340) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117214 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35678 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43591 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/65912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9fa61166-5a82-40fa-b741-6559fa370e20) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118273 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28444 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68006 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21644 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65079 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97829 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21758 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42265 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31648 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/124585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42633 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99711 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24471 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41414 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19268 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42147 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47701 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/41658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/44983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/43383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -559,29 +559,30 @@ static int textAlignResolvingStartAndEnd(T& style)
-    else if (tabSpanNode(node))
-        node = node->parentNode();
+    RefPtr protectedNode = node;
+    if (CheckedPtr tabSpan = parentTabSpanNode(protectedNode.get()))
```
We should use `RefPtr` instead.

rniwa commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1719,10 +1720,10 @@ WritingDirection EditingStyle::textDirectionForSelection(const VisibleSelection&
         end = selection.end().upstream();
-        for (auto& intersectingNode : intersectingNodes(*makeSimpleRange(position, end))) {
-            if (!intersectingNode.isStyledElement())
+        for (CheckedRef intersectingNode : intersectingNodes(*makeSimpleRange(position, end))) {
```
Ditto about using `RefPtr`.

rniwa commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -559,29 +559,30 @@ static int textAlignResolvingStartAndEnd(T& style)
 
     if (propertiesToInclude == PropertiesToInclude::EditingPropertiesInEffect) {
-        if (RefPtr<CSSValue> value = backgroundColorInEffect(node))
+        if (RefPtr<CSSValue> value = backgroundColorInEffect(protectedNode.get()))
```
We can just say `RefPtr` instead of `RefPtr<CSSValue>`.

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -559,29 +559,30 @@ static int textAlignResolvingStartAndEnd(T& style)
-        node = tabSpan->parentNode();
-    else if (tabSpanNode(node))
-        node = node->parentNode();
+    RefPtr protectedNode = node;
```
We should really be protecting at the call sites of `EditingStyle::init` instead of using a local protector

nt1m replied:
(which is everywhere we pass a node to `EditingStyle::create`)

vitorroriz replied:
The problem is that "node" gets reassigned here, so we would assign to a raw pointer

nt1m replied:
We should really not use the `protected` prefix, maybe the argument could be `initialNode` and this local variable `node`.

vitorroriz replied:
This PR is handling specifically the problem of local unchecked variables, not all safe C++ issues at once, more PRs will come.

nt1m replied:
My point is that the `protected` prefix is against our C++ guidelines

rniwa replied:
I think it's okay to use `protectedFoo` in cases like this especially if it's a temporary solution.

nt1m replied:
Didn't Vitor say this was permanent since we re-assign the node?

nt1m replied:
If it's permanent, my suggestion is to use `initialNode` (better names welcome) for the argument name, and just `node` for the local variable.

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -557,21 +557,22 @@ static int textAlignResolvingStartAndEnd(T& style)
 
     if (propertiesToInclude == PropertiesToInclude::EditingPropertiesInEffect) {
-        if (RefPtr<CSSValue> value = backgroundColorInEffect(node))
+        if (RefPtr<CSSValue> value = backgroundColorInEffect(node.get()))
```
```suggestion
        if (RefPtr value = backgroundColorInEffect(node.get()))
```

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -557,21 +557,22 @@ static int textAlignResolvingStartAndEnd(T& style)
-        if (RefPtr<CSSValue> value = backgroundColorInEffect(node))
+        if (RefPtr<CSSValue> value = backgroundColorInEffect(node.get()))
             m_mutableStyle->setProperty(CSSPropertyBackgroundColor, value->cssText(CSS::defaultSerializationContext()));
         if (RefPtr<CSSValue> value = computedStyleAtPosition.propertyValue(CSSPropertyWebkitTextDecorationsInEffect)) {
```
```suggestion
        if (RefPtr value = computedStyleAtPosition.propertyValue(CSSPropertyWebkitTextDecorationsInEffect)) {
```

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -557,21 +557,22 @@ static int textAlignResolvingStartAndEnd(T& style)
 {
-    if (auto* tabSpan = parentTabSpanNode(node))
+    RefPtr node = initialNode;
+    if (CheckedPtr tabSpan = parentTabSpanNode(node.get()))
```
```suggestion
    if (RefPtr tabSpan = parentTabSpanNode(node.get()))
```

vitorroriz committed:
EditingStyle: Avoid unchecked local variables
rdar://157719495
https://bugs.webkit.org/show_bug.cgi?id=297036

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::textDirectionForSelection):

Canonical link: https://commits.webkit.org/298372@main
